### ISSUE TITLE
Possibility to influence word placement

### ISSUE THREAD

**Issue by boidolr** [amueller/word_cloud#293]
I was wondering whether it would be possible to include an option to modify word placement. Currently it is random (only dependant on the number of possible positions):
https://github.com/amueller/word_cloud/blob/af7afbdda1fc4c7e8004da66b3870bd74cab875c/wordcloud/query_integral_image.pyx#L25

With such an option one could for example place connected words close to each other in the resulting image, where 'connected' could be word proximity in the input text or words belonging to some group.

**Comment by amueller** [comment]
Yes, we could provide an interface for that. How would you implement this?
Compute a distance based on word-distance or sentence co-occurence and then do distance-weighted sampling?
I think I would rather have the user compute the distances on their end, so you could to topic models or word2vec or something like that and then call ``generate_from_frequencies`` with a distance metric and the words.
But then you need to sample from the distance-weighted distribution. Is there an easy way to do that? I guess since it's on a grid, we could just evaluate all points and sample from the multinomial. Is that slow?

PR welcome ;)

**Comment by boidolr** [comment]
I am not convinced that accepting a distance metric is the way to go. As you say, evaluating that on every word might be quite expensive (and probably tricky to track). A simpler variant is just letting the client decide its own rule for selecting the position from all available options. That would allow implementations to minimize distance to all previously placed, similar words, or just to the last placed similar word, or any other option.

As I am currently stuck with getting it to work with the C extensions so here is a draft:
```
def query_integral_image(unsigned int[:,:] integral_image, int size_x, int
                         size_y, position_strategy):
    # ... some code here
    candidates = []
    for i in xrange(x - size_x):
        for j in xrange(y - size_y):
            area = integral_image[i, j] + integral_image[i + size_x, j + size_y]
            area -= integral_image[i + size_x, j] + integral_image[i, j + size_y]
            if not area:
                candidates.append((i,j))
    if not len(candidates):
        return None
    return position_strategy(candidates)
```
If speed is a concern, one could use a `int[:,:]` for the result and adapt the size after the loops are finished, forcing the client to work with `np.ndarray`. The actual selection could, of course, still be a bottleneck.

In `wordcloud.py` in the function calling into that code instead of `random_state` one gets `lambda c: c[random_state.randint(0, len(c))])` as `position_strategy`. If that function were stateful (and the currently placed word were handed into `sample_position`) instead of random selection it could keep track of previous and current words and adapt the selection accordingly.

The strategy itself could be exposed similar to the way `color_func` is handled in the API.

**Comment by amueller** [comment]
sure, that's also a possibility. let me know how it works out.

**Comment by mathieuchateau** [comment]
Hello,
Any update on this ?
I am interested too :)
I already assign same color to linked words, but would help a lot to get them close geographically too

**Comment by amueller** [comment]
@mathieuchateau how would you like the words to be placed?

**Comment by mathieuchateau** [comment]
@amueller : try to put words from same color near each other, to form cluster.
Maybe starting from a corner or so and then adding all words of same color, and then do again for next one?

Thanks :)

**Comment by amueller** [comment]
@mathieuchateau That would require a completely different placement algorithm. I'm not planning to work on this but you can feel free to give it a go.
It's also unclear to me how a user would specify this in the interface.

**Comment by amueller** [comment]
If you have a fixed cluster assignment, you can always create one word-cloud per cluster and assemble them later, similar to what was done here: https://www.reddit.com/r/dataisbeautiful/comments/gs4me1/oc_word_cloud_comparison_between_user_comments_on/

You could also create masks for each corner or something like that. You could even do a yin-yang (that actually would be a pretty cool example?).

**Comment by mathieuchateau** [comment]
@amueller : thanks, great idea, it works :)
I had hard time removing margin at each level, but quite great now.

**Comment by amueller** [comment]
Great! Do you wanna post the result?

**Comment by mathieuchateau** [comment]
Here is what it looks like. Sorry for the blur, not sure I am allowed to publish in clear.
Each block is a wordcloud result, then combined together. Issue remaining is horizontal margin between each block. I could remove the vertical ones, not the horizontal one. May not keep using this way, as it's still a bit limited. 
![image](https://user-images.githubusercontent.com/4478521/83939976-5a23d780-a7e1-11ea-9cfc-186ad7f0dc6f.png)

**Comment by amueller** [comment]
Hm you should be able to not have any spaces, I might put together an example for it.

**Comment by abubelinha** [comment]
I do think an interesting variant of this issue when you want to create a wordcloud of geographical names (countries, states, cities, neighborhoods or whatever).

It would be easy to add x,y coordinates to each of those names, so wordcloud can optionally use those x,y values to make decisions on positioning each word.
I have absolutely no idea about how to implement it, though.  I am just asking for the feature  :)

I think this is a much more natural example of the usefulness of forced-positioning (words in a text are probably repeated in different positions, whereas geographical entities have only one position in the map -as far as your list of names does not contain equally named entities, which uses to be the case-).

**Comment by amueller** [comment]
@abubelinha you don't need the word-cloud function any more, right? you can just call ``matplotlib.text`` or ``PIL`` to position the text at that point? The purpose of the wordcloud library is to create the positioning.

**Comment by amueller** [comment]
Ah, sorry, I thought you meant just providing the given positions, I misunderstood.
For that, a physics-based layouting algorithm might be more suitable, but you could also play with distnace-weighted sampling as suggested above.
